docs/examples/bayesian_optimisation.py

```diff
@@ -0,0 +1,737 @@
+# Sampling Notebook](https://docs.jaxgaussianprocesses.com/examples/spatial/). This method
+# is used as it enables us to sample from the posterior in a manner which scales linearly
+# with the number of points sampled, $O(N)$, mitigating the cubic cost associated with
+# drawing exact samples from a GP posterior, $O(N^3)$. It also generates more accurate
+# samples than many other methods for drawing approximate samples from a GP posterior.
+#
+# Note that we also define a `utility_fn` which calls the approximate
+# sample but returns the value returned as a scalar. This is because the `sample_approx`
+# function returns an array of shape $[N, B]$, with $N$ being the number of points within
+# each sample and $B$ being the number of samples drawn. We'll only be drawing (and
+# optimising) one sample at a time, and our optimiser requires the function being
+# optimised to return a scalar output (only querying it at $N=1$ points), so we'll remove the axes from the returned value.
+
+# %%
+approx_sample = opt_posterior.sample_approx(
+    num_samples=1, train_data=D, key=key, num_features=500
+)
+utility_fn = lambda x: approx_sample(x)[0][0]
+
+
+# %% [markdown]
+# In order to minimise the sample, we'll be using the L-BFGS-B ([Byrd et al., 1995](https://epubs.siam.org/doi/abs/10.1137/0916069)) optimiser from the `jaxopt`
+# library. This is a gradient-based optimiser which performs optimisation within a bounded
+# domain. In order to perform optimisation, this optimiser requires a point to start from.
+# Therefore, we will first query our sample from the posterior at a random set of points,
+# and then use the lowest point from this set of points as the starting point for the
+# optimiser. In this example we'll sample 100 points from the posterior, due to the simple
+# nature of the Forrester function. However, in practice it can be beneficial to
+# adopt a more sophisticated approach, and there are several heuristics available in the
+# literature (see for example ([Le Riche and Picheny,
+# 2021](https://arxiv.org/abs/2103.16649))). For instance, one may randomly sample the
+# posterior at a number of points proportional to the dimensionality of the input space,
+# and one may run gradient-based optimisation from multiple of these points, to reduce the
+# risk of converging upon local minima.
+
+
+# %%
+def optimise_sample(
+    sample: FunctionalSample,
+    key: Int[Array, ""],
+    lower_bound: Float[Array, "D"],
+    upper_bound: Float[Array, "D"],
+    num_initial_sample_points: int,
+) -> ScalarFloat:
+    initial_sample_points = jr.uniform(
+        key,
+        shape=(num_initial_sample_points, lower_bound.shape[0]),
+        dtype=jnp.float64,
+        minval=lower_bound,
+        maxval=upper_bound,
+    )
+    initial_sample_y = sample(initial_sample_points)
+    best_x = jnp.array([initial_sample_points[jnp.argmin(initial_sample_y)]])
+
+    negative_utility_fn = lambda x: sample(x)[0][0]
```

Comment:
maybe its worth stressing that we use negative_utility_fn because the optimizer is a minimiser?